### ISSUE TITLE
Problem on restoring the policy after training

### ISSUE THREAD

**Issue by lsylusiyao** [aai-institute/tianshou#724]
I have trained my policy until some good results for both training and testing. Everything works fine when I pass the `policy` object to another function and test it. However, when I terminated the program and restored the policy again, no matter environment from the training and testing or the environment in my own function failed on the policy as if it hadn't been trained.  Here's part of my code for training and restoring. 

```python
# wrapper is the action mask
lambda_envs = [lambda: Wrapper(MyEnv(xxx)) for _ in range(10)]
train_envs = SubprocVectorEnv(lambda_envs)
lambda_envs = [lambda: Wrapper(MyEnv(xxx)) for _ in range(4)]
test_envs = SubprocVectorEnv(lambda_envs)

Q_param = V_param = {"hidden_sizes": [128]}

net = Net(
        obs_shape,
        action_shape,
        hidden_sizes=config["hidden-sizes"],
        device=device,
        softmax=True,
        num_atoms=config["num-atoms"],
        dueling_param=(Q_param, V_param)
).to(device)
optim = torch.optim.Adam(net.parameters(), lr=config["lr"])
policy = RainbowPolicy(
        net,
        optim,
        config["gamma"],
        config["num-atoms"],
        config["v-min"],
        config["v-max"],
        config["n-step"],
        target_update_freq=config["target-update-freq"]
    ).to(device)

buf = PrioritizedVectorReplayBuffer(
            config["buffer-size"],
            buffer_num=len(train_envs),
            alpha=config["alpha"],
            beta=config["beta"],
            weight_norm=True
        )

train_collector = Collector(policy, train_envs, buf, exploration_noise=True)
test_collector = Collector(policy, test_envs, exploration_noise=True)
train_collector.collect(n_step=config['batch-size'] * len(train_envs))

logger = TensorboardLogger(writer)
if resume:
        # load from existing checkpoint
        print(f"Loading agent under {restore_dir}")
        ckpt_path = os.path.join(restore_dir, 'checkpoint.pth')
        if os.path.exists(ckpt_path):
            checkpoint = torch.load(ckpt_path, map_location=device)
            policy.load_state_dict(checkpoint['model'])
            policy.optim.load_state_dict(checkpoint['optim'])
            policy.optim.param_groups[0]['capturable'] = True
            print("Successfully restore policy and optim.")
        else:
            print("Fail to restore policy and optim.")
        buffer_path = os.path.join(restore_dir, 'train_buffer.pkl')
        if os.path.exists(buffer_path):
            train_collector.buffer = pickle.load(open(buffer_path, "rb"))
            print("Successfully restore buffer.")
        else:
            print("Fail to restore buffer.")

result = offpolicy_trainer(
        policy,
        train_collector,
        test_collector,
        epoch,
        config["step-per-epoch"],
        config["step-per-collect"],
        config["test-count"],
        config["batch-size"],
        update_per_step=config["update-per-step"],
        train_fn=train_fn,
        test_fn=lambda epoch, env_step: policy.set_eps(config["eps-test"]),
        save_best_fn=lambda policy_temp: torch.save(policy_temp.state_dict(), 
            os.path.join(log_path, "policy_best.pth")),
        logger=logger,
        resume_from_log=resume,
        save_checkpoint_fn=save_checkpoint_fn
    )
```

`resume -> True` when I tried to resume the policy. And here is my own function.

```python
from tianshou.data.batch import Batch
agent.eval()
env = MyEnv()
obs = env.reset()
done = False
while not done:
    action = agent.forward(Batch(
        obs = Batch(
            obs = np.array([obs], dtype=np.float32),
            mask=env.action_states()),
        info = Batch())).act
    action = int(action)
    obs, reward, done, info = env.step(action)
    steps += 1
    if steps > 1e2:
        print("Test Failed!!!\n")
        break

```

This function works no matter whether I use the action mask or not just after training. But it fails after the restore or resume.

**Comment by Trinkle23897** [comment]
Have you ever printed the numerical result of one layer of the trained network and the restored network?

For quick debugging, you can train it for one epoch and kill the training process, then reload the model weight to see if there's any difference.

**Comment by lsylusiyao** [comment]
Well, I've tested the results from training and reloading for both 1 env and training 1 time, or 4 env and training 10 times. The print results are always equal, and my validating environment can perform the same as the testing environment. Even the outputs from the agents are the same. It seems that it has nothing to do with our framework. I'll recheck my code. Thanks.
